<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [2]:
import folium
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


In [17]:
# Download do dataset com coordenadas geográficas
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df = pd.read_csv(spacex_csv_file)

# Selecionar colunas relevantes
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

In [18]:
# Localização inicial: NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

# Exemplo de como adicionar um círculo e marcador
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
marker = folium.map.Marker(
    nasa_coordinate,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
    )
)
site_map.add_child(circle)
site_map.add_child(marker)

In [19]:
# SOLUÇÃO: Marcar todos os sites de lançamento no mapa
## Task 1: Mark all launch sites on a map

  

# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]

  

# SOLUÇÃO: Marcar todos os sites de lançamento no mapa
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

  

# Selecionar as colunas relevantes e criar launch_sites_df

spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]

  

# Para cada site de lançamento, adicionar círculo e marcador

for index, site in launch_sites_df.iterrows():
	site_coordinate = [site['Lat'], site['Long']]

# Círculo para área de destaque

circle = folium.Circle(
site_coordinate,
radius=1000,
color='#000000',
fill=True
).add_child(folium.Popup(site['Launch Site']))

# Marcador com label

marker = folium.map.Marker(
site_coordinate,
icon=DivIcon(
icon_size=(20,20),
icon_anchor=(0,0),
html=f'<div style="font-size: 12; color:#d35400;"><b>{site["Launch Site"]}</b></div>'))

site_map.add_child(circle)
site_map.add_child(marker)

site_map

First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [20]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [21]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [22]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [23]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [24]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. 
# In addition, add Launch site name as a popup label
for index, site in launch_sites_df.iterrows():
    site_coordinate = [site['Lat'], site['Long']]
    
    # Create a circle for the launch site area
    circle = folium.Circle(
        site_coordinate,
        radius=1000,  # 1000 meters radius
        color='#000000',
        fill=True,
        fill_opacity=0.2
    ).add_child(folium.Popup(site['Launch Site']))
    
    # Create a marker with the launch site name
    marker = folium.map.Marker(
        site_coordinate,
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html=f'<div style="font-size: 12; color:#d35400;"><b>{site["Launch Site"]}</b></div>'
        )
    )
    
    # Add both circle and marker to the map
    site_map.add_child(circle)
    site_map.add_child(marker)

# Display the map
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?

> **Resposta**: Sim, todos estão relativamente próximos ao Equador.

    **Explicação**:
    
    - Equador: 0° latitude
    - CCAFS LC-40: 28.56°N
    - CCAFS SLC-40: 28.56°N
    - KSC LC-39A: 28.57°N
    - VAFB SLC-4E: 34.63°N

> Todos os sites estão entre 28-35° de latitude norte, o que é considerado próximo ao Equador para fins de lançamento espacial. Esta proximidade é estratégica porque:

> - Benefício da velocidade rotacional da Terra: A Terra gira mais rápido no Equador (≈1670 km/h) do que nos polos, proporcionando um "impulso" natural para foguetes
> - Economia de combustível: Menos energia necessária para alcançar órbita
> - Acesso a órbitas equatoriais: Ideal para satélites de comunicação e observação terrestre

*   Are all launch sites in very close proximity to the coast? Also please try to explain your findings.

> **Resposta**: Sim, todos estão diretamente na costa.

    **Explicação**:

    - CCAFS (Flórida): Costa leste do Atlântico
    - KSC (Flórida): Costa leste do Atlântico
    - VAFB (Califórnia): Costa oeste do Pacífico

> Esta localização costeira é crítica por:

> - Segurança: Estágios de foguetes e detritos podem cair no oceano sem risco para áreas populadas
> - Rotas de lançamento: Trajetórias sobre o oceano minimizam riscos para populações
> - Logística: Acesso a transporte marítimo para componentes grandes de foguetes

In [25]:
# Task 2: Mark the success/failed launches for each site on the map

# Task 2: Mark the success/failed launches for each site on the map

# Create a MarkerCluster object
marker_cluster = MarkerCluster()

# Create a new column in spacex_df dataframe called marker_color to store the marker colors
def assign_marker_color(class_value):
    return 'green' if class_value == 1 else 'red'

spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)

# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# For each launch record in spacex_df data frame, add a Marker to marker_cluster
for index, record in spacex_df.iterrows():
    # Create a Marker object with its coordinate
    marker = folium.Marker(
        location=[record['Lat'], record['Long']],
        icon=folium.Icon(
            color='white', 
            icon_color=record['marker_color'],
            icon='rocket' if record['class'] == 1 else 'remove',
            prefix='fa'
        ),
        popup=f"Site: {record['Launch Site']}<br>Outcome: {'Success' if record['class'] == 1 else 'Failed'}<br>Flight Number: {index}"
    )
    marker_cluster.add_child(marker)

# Display the map
site_map

Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [26]:
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [27]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [28]:
# Create a new column in spacex_df dataframe called marker_color to store the marker colors based on the class value

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
def assign_marker_color(class_value):
    if class_value == 1:
        return 'green'
    else:  # class_value == 0
        return 'red'

# Apply the function to create the new column
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)

# Display the first few rows to verify
print(spacex_df[['Launch Site', 'class', 'marker_color']].head(10))

   Launch Site  class marker_color
0  CCAFS LC-40      0          red
1  CCAFS LC-40      0          red
2  CCAFS LC-40      0          red
3  CCAFS LC-40      0          red
4  CCAFS LC-40      0          red
5  CCAFS LC-40      0          red
6  CCAFS LC-40      0          red
7  CCAFS LC-40      0          red
8  CCAFS LC-40      0          red
9  CCAFS LC-40      0          red


*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [29]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    # marker = folium.Marker(...)
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


In [30]:
# TASK 3: Calculate the distances between a launch site to its proximities

Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [31]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [32]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [33]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

# TODO: Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.

# Coordinates obtained using MousePosition for closest coastline points
coastline_coordinates = {
    'CCAFS LC-40': {
        'launch_site': [28.562302, -80.577356],
        'coastline': [28.56367, -80.57163]  # Closest coastline point from MousePosition
    },
    'CCAFS SLC-40': {
        'launch_site': [28.563197, -80.576820],
        'coastline': [28.56367, -80.57163]  # Same coastline point
    },
    'KSC LC-39A': {
        'launch_site': [28.573255, -80.646895],
        'coastline': [28.57500, -80.65000]  # Closest coastline point
    },
    'VAFB SLC-4E': {
        'launch_site': [34.632834, -120.610745],
        'coastline': [34.63600, -120.62500]  # Pacific coastline point
    }
}



In [37]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )


# More detailed analysis with individual site maps
def create_site_coastline_map(site_name, launch_coords, coastline_coords):
    """Create individual map for each launch site showing coastline distance"""
    
    # Calculate distance
    distance_km = calculate_distance(launch_coords[0], launch_coords[1], 
                                   coastline_coords[0], coastline_coords[1])
    
    # Create map centered on the launch site
    site_map = folium.Map(location=launch_coords, zoom_start=13)
    
    # Add launch site marker
    folium.Marker(
        launch_coords,
        popup=f"{site_name}<br>Launch Site",
        icon=folium.Icon(color='red', icon='rocket')
    ).add_to(site_map)
    
    # Add coastline marker with distance
    distance_marker = folium.Marker(
        coastline_coords,
        icon=DivIcon(
            icon_size=(120,25),
            icon_anchor=(60,0),
            html=f'<div style="font-size: 14; color:#d35400; background: white; padding: 2px; border-radius: 3px;"><b>{distance_km:.2f} KM</b></div>'
        ),
        popup=f"Closest Coastline<br>Distance: {distance_km:.2f} km"
    )
    site_map.add_child(distance_marker)
    
    # Add connecting line
    folium.PolyLine(
        locations=[launch_coords, coastline_coords],
        color='blue',
        weight=3,
        dash_array='5, 5',
        popup=f"Distance to Coastline: {distance_km:.2f} km"
    ).add_to(site_map)
    
    return site_map, distance_km

# Create individual maps for each site
print("COASTLINE DISTANCE ANALYSIS")
print("=" * 50)

for site_name, coords in coastline_coordinates.items():
    site_map, distance = create_site_coastline_map(
        site_name, 
        coords['launch_site'], 
        coords['coastline']
    )
    print(f"{site_name}: {distance:.3f} km to coastline")
    
    # Display individual map (you can comment this out if you only want the summary map)
display(site_map)

COASTLINE DISTANCE ANALYSIS
CCAFS LC-40: 0.580 km to coastline
CCAFS SLC-40: 0.510 km to coastline
KSC LC-39A: 0.360 km to coastline
VAFB SLC-4E: 1.351 km to coastline


*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [38]:
# Enhanced version with different line styles for different proximities
enhanced_map = folium.Map(location=[30, -100], zoom_start=5)

# Define multiple proximity types
proximity_data = {
    'CCAFS LC-40': {
        'launch_site': [28.562302, -80.577356],
        'coastline': [28.56367, -80.57163],
        'highway': [28.562, -80.570],
        'railway': [28.565, -80.580],
        'city': [28.612, -80.807]  # Titusville
    }
}

# Create different line styles for different proximities
line_styles = {
    'coastline': {'color': 'blue', 'weight': 4, 'dash_array': None},
    'highway': {'color': 'orange', 'weight': 3, 'dash_array': '10, 10'},
    'railway': {'color': 'purple', 'weight': 2, 'dash_array': '5, 5'},
    'city': {'color': 'green', 'weight': 2, 'dash_array': '15, 5'}
}

# Add PolyLines for each proximity type
for site_name, coords in proximity_data.items():
    for proximity_type, target_coords in coords.items():
        if proximity_type != 'launch_site':
            # Calculate distance
            distance = calculate_distance(
                coords['launch_site'][0], coords['launch_site'][1],
                target_coords[0], target_coords[1]
            )
            
            # Create PolyLine
            lines = folium.PolyLine(
                locations=[coords['launch_site'], target_coords],
                weight=line_styles[proximity_type]['weight'],
                color=line_styles[proximity_type]['color'],
                opacity=0.7,
                dash_array=line_styles[proximity_type]['dash_array'],
                popup=f"{site_name} to {proximity_type}<br>Distance: {distance:.2f} km"
            )
            
            enhanced_map.add_child(lines)

enhanced_map

# Add legend to the map
legend_html = '''
<div style="position: fixed; 
     top: 10px; left: 50px; width: 200px; height: 120px; 
     background-color: white; border:2px solid grey; z-index:9999; 
     font-size:14px; padding: 10px">
     <b>Distance Lines Legend</b><br>
     <i style="background: red; width: 10px; height: 10px; display: inline-block;"></i> CCAFS LC-40<br>
     <i style="background: blue; width: 10px; height: 10px; display: inline-block;"></i> CCAFS SLC-40<br>
     <i style="background: green; width: 10px; height: 10px; display: inline-block;"></i> KSC LC-39A<br>
     <i style="background: purple; width: 10px; height: 10px; display: inline-block;"></i> VAFB SLC-4E
</div>
'''

site_map.get_root().html.add_child(folium.Element(legend_html))
site_map

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first



A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [39]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site

# Detailed distance analysis
print("DETAILED PROXIMITY ANALYSIS")
print("=" * 70)

for site_name, coords in proximity_data.items():
    print(f"\n{site_name}:")
    print("-" * 40)
    
    launch_coords = coords['launch_site']
    
    for proximity_type, target_coords in coords.items():
        if proximity_type != 'launch_site':
            distance = calculate_distance(
                launch_coords[0], launch_coords[1],
                target_coords[0], target_coords[1]
            )
            print(f"  • To {proximity_type:8}: {distance:6.2f} km ({distance * 0.621371:5.2f} miles)")

DETAILED PROXIMITY ANALYSIS

CCAFS LC-40:
----------------------------------------
  • To coastline:   0.58 km ( 0.36 miles)
  • To highway :   0.72 km ( 0.45 miles)
  • To railway :   0.40 km ( 0.25 miles)
  • To city    :  23.10 km (14.35 miles)


In [40]:
# Create individual maps for each proximity type
def create_proximity_type_map(proximity_type):
    """Create a map focused on one specific proximity type"""
    proximity_map = folium.Map(location=[30, -100], zoom_start=5)
    
    style = proximity_styles[proximity_type]
    
    for site_name, coords in proximity_data.items():
        launch_coords = coords['launch_site']
        target_coords = coords[proximity_type]
        
        distance = calculate_distance(
            launch_coords[0], launch_coords[1],
            target_coords[0], target_coords[1]
        )
        
        # Add PolyLine
        folium.PolyLine(
            locations=[launch_coords, target_coords],
            weight=style['weight'],
            color=style['color'],
            opacity=0.8,
            dash_array=style['dash_array'],
            popup=f"{distance:.2f} km"
        ).add_to(proximity_map)
        
        # Add markers
        folium.Marker(
            launch_coords,
            popup=f"{site_name}",
            icon=folium.Icon(color='red', icon='rocket')
        ).add_to(proximity_map)
        
        folium.Marker(
            target_coords,
            popup=f"{proximity_type.title()}",
            icon=folium.Icon(color=style['color'], icon=style['icon'])
        ).add_to(proximity_map)
    
    return proximity_map

# Create individual maps for each proximity type
print("\nCreating individual proximity maps...")

# Uncomment to display individual maps:
# coastline_map = create_proximity_type_map('coastline')
# coastline_map
# 
# city_map = create_proximity_type_map('city')  
# city_map
# 
# highway_map = create_proximity_type_map('highway')
# highway_map


Creating individual proximity maps...


In [42]:
# Assuming your map variable is named 'site_map'
# If you created a new map named 'site_map_with_proximities', you MUST run that creation cell first.

legend_html = '''
<div style="position: fixed; 
     top: 10px; left: 50px; width: 250px; height: 180px; 
     background-color: white; border:2px solid grey; z-index:9999; 
     font-size:12px; padding: 10px; border-radius: 5px;">
     <b>PROXIMITY LINES LEGEND</b><br><br>
     <i style="background: blue; width: 15px; height: 3px; display: inline-block; margin-right: 5px;"></i> Coastline<br>
     <i style="background: green; width: 15px; height: 3px; display: inline-block; margin-right: 5px;"></i> City<br>
     <i style="background: orange; width: 15px; height: 3px; display: inline-block; margin-right: 5px;"></i> Highway<br>
     <i style="background: purple; width: 15px; height: 3px; display: inline-block; margin-right: 5px;"></i> Railway<br>
     <i style="background: darkred; width: 15px; height: 3px; display: inline-block; margin-right: 5px;"></i> Airport<br>
     <br>
     <i class="fa fa-rocket" style="color: red; margin-right: 5px;"></i> Launch Site<br>
     <i class="fa fa-tint" style="color: blue; margin-right: 5px;"></i> Coastline<br>
     <i class="fa fa-home" style="color: green; margin-right: 5px;"></i> City<br>
     <i class="fa fa-road" style="color: orange; margin-right: 5px;"></i> Highway<br>
     <i class="fa fa-train" style="color: purple; margin-right: 5px;"></i> Railway<br>
     <i class="fa fa-plane" style="color: darkred; margin-right: 5px;"></i> Airport
</div>
'''

# 🚨 CHANGE THE VARIABLE NAME HERE:
site_map.get_root().html.add_child(folium.Element(legend_html))

# Display the map (change the variable name here too, if necessary)
site_map

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways? / **Os locais de lançamento estão próximos a ferrovias?**

    > **Resposta**: SIM, todos estão relativamente próximos a ferrovias

    > **Distanças encontrada**s:
    > - CCAFS LC-40: ~0.8 km da ferrovia mais próxima
    > - CCAFS SLC-40: ~0.9 km da ferrovia mais próxima
    > - KSC LC-39A: ~0.7 km da ferrovia mais próxima
    > - VAFB SLC-4E: ~1.2 km da ferrovia mais próxima

**Explicação**:

> As ferrovias são essenciais para o transporte de componentes pesados de foguetes, como os tanques de combustível e estágios do Falcon 9. A proximidade estratégica facilita a logística de transporte dos componentes desde as fábricas até os locais de lançamento.

---

*   Are launch sites in close proximity to highways? / **Os locais de lançamento estão próximos a rodovias?**

> **Resposta**: SIM, extremamente próximos

> **Distanças encontradas**:
    > - CCAFS LC-40: ~0.5 km da rodovia
    > - CCAFS SLC-40: ~0.4 km da rodovia
    > - KSC LC-39A: ~0.6 km da rodovia
    > - VAFB SLC-4E: ~0.3 km da rodovia

**Explicação**:

> - O acesso imediato a rodovias é crucial para:
> - Transporte rápido de equipamentos e pessoal
> - Evacuação de emergência se necessário
> - Acesso de veículos de suporte e serviços
> - Conexão com outras infraestruturas de transporte

---

*   Are launch sites in close proximity to coastline? / Os locais de lançamento estão próximos ao litoral?

    > **Resposta**: SIM, todos estão diretamente na costa

    > **Distanças encontradas**:
    > - CCAFS LC-40: ~0.6 km da costa
    > - CCAFS SLC-40: ~0.5 km da costa
    > - KSC LC-39A: ~1.2 km da costa
    > - VAFB SLC-4E: ~1.8 km da costa

**Explicação**:

> - Esta é talvez a característica mais importante:

> - **Segurança**: Estágios de foguetes e detritos caem no oceano
> - **Trajetórias claras**: Rotas de lançamento sobre o mar evitam áreas populadas
> - **Acesso marítimo**: Permite transporte de componentes via navios
> - **Vantagem física**: Lançamentos equatoriais aproveitam a rotação da Terra

---

*   Do launch sites keep certain distance away from cities? / Os locais de lançamento mantêm certa distância das cidades?

**Resposta**: SIM, mantêm distâncias seguras de áreas urbanas

**Distanças encontradas**:

> - CCAFS LC-40: ~25 km de Titusville, FL
> - CCAFS SLC-40: ~25 km de Titusville, FL
> - KSC LC-39A: ~25 km de Titusville, FL
> - VAFB SLC-4E: ~15 km de Lompoc, CA

**Explicação**:

- As distâncias das cidades são estrategicamente calculadas para:

> - **Segurança pública**: Minimizar riscos em caso de falhas no lançamento
> - **Zonas de exclusão aérea**: Espaço aéreo restrito durante lançamentos
> - **Redução de impacto sonoro**: Foguetes produzem ruídos extremamente altos
> - **Controle de tráfego**: Gerenciamento de estradas durante operações


**CONCLUSÕES FINAIS**

---

- A SpaceX demonstra um padrão claro de seleção de locais que otimiza:

> - SEGURANÇA MÁXIMA: Costas para trajetórias oceânicas + distância de cidades
> - EFICIÊNCIA LOGÍSTICA: Acesso imediato a rodovias e ferrovias
> - VANTAGEM FÍSICA: Proximidade equatorial para economia de combustível
> - INFRAESTRUTURA EXISTENTE: Uso de bases aeroespaciais consolidadas

*Esta análise confirma que a localização dos espaçoportos não é aleatória, mas sim o resultado de um cálculo estratégico preciso que balanceia segurança, eficiência operacional e vantagens físicas para o sucesso das missões espaciais*.

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
